lib/Driver/ToolChains.cpp

```diff
@@ -1471,6 +1519,222 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
   return II;
 }
 
+ToolChain::InvocationInfo
+toolchains::Windows::constructInvocation(const InterpretJobAction &job,
+                                             const JobContext &context) const {
+  InvocationInfo II = ToolChain::constructInvocation(job, context);
+
+  return II;
+}
+
+ToolChain::InvocationInfo
+toolchains::Windows::constructInvocation(const LinkJobAction &job,
+                                             const JobContext &context) const {
+  assert(context.Output.getPrimaryOutputType() == types::TY_Image &&
+         "Invalid linker output type.");
+
+  ArgStringList Arguments;
+
+  switch (job.getKind()) {
+  case LinkKind::None:
+    llvm_unreachable("invalid link kind");
+  case LinkKind::Executable:
+    // Default case, nothing extra needed.
+    break;
+  case LinkKind::DynamicLibrary:
+    Arguments.push_back("-shared");
+    break;
+  }
+
+  // Select the linker to use.
+  std::string Linker;
+  if (const Arg *A = context.Args.getLastArg(options::OPT_use_ld)) {
+    Linker = A->getValue();
+  }
+  if (!Linker.empty()) {
+    Arguments.push_back(context.Args.MakeArgString("-fuse-ld=" + Linker));
+  }
+
+  // Configure the toolchain.
+  // By default, use the system clang++ to link.
+  const char * Clang = "clang++";
```

Comment:
Has there been any discussion about using `cl` directly? Do we get any benefit from using Clang?

Reply:
I think there are tradeoffs to each approach. With `cl`, the user doesn't need Clang installed, but Clang does provide things like ASAN that MSVC is missing. If we also might end up with a dependency on Compiler-RT (see https://github.com/apple/swift/pull/13234) then it'd probably be worth sticking with Clang.

As for using the `clang-cl` vs. `clang` driver: I don't have any strong opinions there. I suppose there'd be some benefit in consistency with other platforms, but if the consensus is that `clang-cl` is preferred I can easily switch over the commands.

Reply:
I suppose one option would be to use `clang-cl` by default and fallback to `cl` if Clang isn't installed.